airflow/models/taskinstance.py

```diff
@@ -1265,6 +1558,132 @@ def _update_rtif(ti, rendered_fields, session: Session | None = None):
     RenderedTaskInstanceFields.delete_old_records(ti.task_id, ti.dag_id, session=session)
 
 
+def _coalesce_to_orm_ti(*, ti: TaskInstancePydantic | TaskInstance, session: Session):
+    from airflow.models.dagrun import DagRun
+    from airflow.serialization.pydantic.taskinstance import TaskInstancePydantic
+
+    if isinstance(ti, TaskInstancePydantic):
+        orm_ti = DagRun.fetch_task_instance(
+            dag_id=ti.dag_id,
+            dag_run_id=ti.run_id,
+            task_id=ti.task_id,
+            map_index=ti.map_index,
+            session=session,
+        )
+        if TYPE_CHECKING:
+            assert orm_ti
+        ti, pydantic_ti = orm_ti, ti
+        _set_ti_attrs(ti, pydantic_ti)
+        ti.task = pydantic_ti.task
+    return ti
+
+
+@internal_api_call
+@provide_session
+def _defer_task(
+    ti: TaskInstance | TaskInstancePydantic, exception: TaskDeferred, session: Session = NEW_SESSION
+) -> TaskInstancePydantic | TaskInstance:
+    from airflow.models.trigger import Trigger
+
+    # First, make the trigger entry
+    trigger_row = Trigger.from_object(exception.trigger)
+    session.add(trigger_row)
+    session.flush()
+
+    ti = _coalesce_to_orm_ti(ti=ti, session=session)  # ensure orm obj in case it's pydantic
+
+    if TYPE_CHECKING:
+        assert ti.task
+
+    # Then, update ourselves so it matches the deferral request
+    # Keep an eye on the logic in `check_and_change_state_before_execution()`
+    # depending on self.next_method semantics
+    ti.state = TaskInstanceState.DEFERRED
+    ti.trigger_id = trigger_row.id
+    ti.next_method = exception.method_name
+    ti.next_kwargs = exception.kwargs or {}
+
+    # Calculate timeout too if it was passed
+    if exception.timeout is not None:
+        ti.trigger_timeout = timezone.utcnow() + exception.timeout
+    else:
+        ti.trigger_timeout = None
+
+    # If an execution_timeout is set, set the timeout to the minimum of
+    # it and the trigger timeout
+    execution_timeout = ti.task.execution_timeout
+    if execution_timeout:
+        if TYPE_CHECKING:
+            assert ti.start_date
+        if ti.trigger_timeout:
+            ti.trigger_timeout = min(ti.start_date + execution_timeout, ti.trigger_timeout)
+        else:
+            ti.trigger_timeout = ti.start_date + execution_timeout
+    if ti.test_mode:
+        _add_log(event=ti.state, task_instance=ti, session=session)
+    session.merge(ti)
+    session.commit()
+    return ti
+
+
+@internal_api_call
+@provide_session
+def _handle_reschedule(
```

Comment:
as above - make naming distinct.
```suggestion
def _handle_ti_reschedule(
```